src/Microsoft.OData.Core/UriParser/Parsers/FunctionOverloadResolver.cs

```diff
@@ -86,24 +87,24 @@ internal static bool ResolveOperationImportFromList(string identifier, IList<str
             }
 
             // If parameter count is zero and there is one function import whoese parameter count is zero, return this function import.
-            if (candidateMatchingOperationImports.Count > 1 && parameterNames.Count == 0)
+            if (candidateMatchingOperationImports.Count() > 1 && parameterNames.Count == 0)
             {
-                candidateMatchingOperationImports = candidateMatchingOperationImports.Where(operationImport => operationImport.Operation.Parameters.Count() == 0).ToList();
+                candidateMatchingOperationImports = candidateMatchingOperationImports.Where(operationImport => operationImport.Operation.Parameters.Count() == 0);
             }
 
-            if (candidateMatchingOperationImports.Count == 0)
+            if (candidateMatchingOperationImports.Count() == 0)

```

Comment:
>.Count() [](start = 49, length = 8)

is it cheaper to use FirstOrDefault() == null?

Reply:
From the Count() function:

```
public static int Count<TSource>(this IEnumerable<TSource> source) {
            if (source == null) throw Error.ArgumentNull("source");
            ICollection<TSource> collectionoft = source as ICollection<TSource>;
            if (collectionoft != null) return collectionoft.Count;
            ICollection collection = source as ICollection;
            if (collection != null) return collection.Count;
            int count = 0;
            using (IEnumerator<TSource> e = source.GetEnumerator()) {
                checked {
                    while (e.MoveNext()) count++;
                }
            }
            return count;
        }
```

So, if the source is ICollection< T >, it's cheaper than "FirstOrDefault< T >". In our codes, internally the IEnumerable< T > is the IList< T >, which implements the ICollection< T >, So, here the Count() is cheaper than "FirstOrDefault< T >".



---
In reply to: [290001862](https://github.com/OData/odata.net/pull/1473#discussion_r290001862) [](ancestors = 290001862)

Reply:
Why not use your new HasAny() extension method?

---
In reply to: [290006737](https://github.com/OData/odata.net/pull/1473#discussion_r290006737) [](ancestors = 290006737,290001862)

Reply:
That's internal extension method defined in OData.Edm library. 

---
In reply to: [290047580](https://github.com/OData/odata.net/pull/1473#discussion_r290047580) [](ancestors = 290047580,290006737,290001862)

Reply:
Should we have a similar method here? For the case where the enumerator is not a collection, this method will actually count all of the members when all we really need to know is if there are any members. I realize that our implementation uses collection, but your code would be more efficient for other enumerators.

---
In reply to: [290048582](https://github.com/OData/odata.net/pull/1473#discussion_r290048582) [](ancestors = 290048582,290047580,290006737,290001862)

Reply:
Got it.

---
In reply to: [290049890](https://github.com/OData/odata.net/pull/1473#discussion_r290049890) [](ancestors = 290049890,290048582,290047580,290006737,290001862)